File: src/adapters/copilot.ts
import { fetchWithRetry, requireField } from "../http-client.js";
import {
  AuthenticationError,
  DataUnavailableError,
  InvalidCliArgError,
} from "../errors.js";
import { sumCost } from "../schema.js";
import type { AdapterResult, DateWindow, UserUsage } from "../schema.js";

const TOOL = "copilot";
const COPILOT_API_VERSION = "2026-03-10";

/**
 * GitHub's Copilot metrics reporting has no data before this date for either
 * orgs or enterprises. A requested window starting earlier throws
 * DataUnavailableError without calling the API at all, mirroring
 * claude-code.ts's ANALYTICS_API_START_DATE guard.
 *
 * Source: GitHub Docs, "Data available in Copilot usage metrics" --
 * enterprise/org usage-metrics reports are available starting 2025-10-10.
 */
const COPILOT_METRICS_START_DATE = "2025-10-10";

/**
 * GitHub's own published, fixed conversion rate from an AI credit to USD --
 * not a negotiated or per-org price, the same $0.01/credit rate applies to
 * every Copilot Business/Enterprise org. Source: GitHub Blog, "GitHub
 * Copilot is moving to usage-based billing" (2026) -- "1 AI credit = $0.01
 * USD". Used to turn the `ai_credits_used` field GitHub's metrics API
 * actually returns into a real dollar figure, rather than inventing one.
 */
const COPILOT_CREDIT_USD_RATE = 0.01;

interface CopilotReportIndex {
  download_links: string[];
  report_day?: string;
}

interface CopilotApiUser {
  user_id: string | number;
  user_login: string;
  ai_credits_used: number;
  user_initiated_interaction_count: number;
}

function splitIntoDays(window: DateWindow): string[] {
  const start = new Date(window.start);
  const end = new Date(window.end);
  const days: string[] = [];
  let cursor = start;

  while (cursor <= end) {
    days.push(cursor.toISOString().slice(0, 10));
    cursor = new Date(cursor);
    cursor.setDate(cursor.getDate() + 1);
  }

  return days;
}

/**
 * Per-day, credits-only usage cost for one user. Does NOT include any
 * seat/license price -- that is added once per user, after merging across
 * every day in the window, by the caller (fetchCopilotSpend), so a
 * multi-day window never double- or triple-counts a flat monthly seat fee.
 */
function normalizeUser(raw: CopilotApiUser): UserUsage {
  const asRecord = raw as unknown as Record<string, unknown>;
  const userLogin = requireField<string>(asRecord, "user_login", TOOL);
  const aiCreditsUsed = requireField<number>(
    asRecord,
    "ai_credits_used",
    TOOL,
  );
  const interactionCount = requireField<number>(
    asRecord,
    "user_initiated_interaction_count",
    TOOL,
  );

  return {
    userId: userLogin,
    // GitHub's Copilot metrics API identifies users by user_id/user_login
    // (a GitHub username), never an email address -- unlike Cursor and
    // Claude Code, there is no email field anywhere in this response.
    userEmail: null,
    // Copilot's metrics reports carry no token counts (input/output/cache)
    // at all -- that concept doesn't apply to its per-user report shape,
    // which is credits- and interaction-count-based instead.
    inputTokens: null,
    outputTokens: null,
    cacheReadTokens: null,
    cacheWriteTokens: null,
    requests: interactionCount,
    costUsd: aiCreditsUsed * COPILOT_CREDIT_USD_RATE,
    // Always true: see fetchCopilotSpend's isEstimated comment below.
    isEstimated: true,
  };
}

function parseNdjson(text: string): CopilotApiUser[] {
  return text
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0)
    .map((line) => JSON.parse(line) as CopilotApiUser);
}

/**
 * Fetches GitHub Copilot org usage for the given window and converts it to a
 * dollar figure, via the CURRENT (non-deprecated) Copilot usage metrics
 * reports API: GET /orgs/{org}/copilot/metrics/reports/users-1-day. The
 * older /orgs/{org}/copilot/metrics endpoint (which returned inline
 * org-aggregate JSON) was sunset by GitHub on 2026-04-02 and no longer
 * exists.
 *
 * Unlike Cursor and Claude Code, this real API has no arbitrary start/end
 * range parameter and no per-user cost field at all:
 *
 * - It only exposes single-day (`users-1-day?day=YYYY-MM-DD`) or
 *   latest-rolling-28-day (`users-28-day/latest`) granularity. To honor an
 *   arbitrary caller-supplied window, this adapter requests one report per
 *   calendar day in the window and sums per-user totals across days --
 *   the same "chunk the window, sum the chunks" shape as
 *   fetchCursorSpend's 30-day pagination, just chunked by day instead.
 * - Each report call doesn't return data inline. It returns
 *   `{ download_links, report_day }`, where download_links are short-lived,
 *   pre-signed GitHub-owned URLs pointing to NDJSON (newline-delimited
 *   JSON) files -- one JSON user record per line, not a single JSON array.
 *   This adapter fetches and parses each one.
 * - Copilot Business/Enterprise is seat-based billing ($19 or $39/seat/
 *   month, bundling a matching monthly AI-credit allowance) -- GitHub does
 *   not expose an org's actual per-seat contract price via any API, the
 *   same structural gap Cursor's and Claude Code's flat-seat plans have.
 *   The only real, vendor-reported-and-attributable number is
 *   `ai_credits_used`, which this adapter converts to USD at GitHub's own
 *   published, fixed $0.01/credit rate (COPILOT_CREDIT_USD_RATE) -- never
 *   fabricated. If `seatPriceUsd` is supplied, it's added once per user
 *   (not once per day) to also reflect the flat license cost; if omitted,
 *   the reported cost is credits-usage-only and explicitly excludes the
 *   seat fee (see README for the same caveat spelled out for CLI users).
 *
 * `isEstimated` is unconditionally true for every Copilot result, regardless
 * of whether seatPriceUsd was supplied: GitHub's Copilot metrics API has no
 * cost_usd/spend_usd-equivalent field at all (unlike Cursor/Claude Code,
 * which do report a native cost figure that is only sometimes suspicious),
 * so every dollar amount this adapter produces is derived, never vendor-
 * reported, and should never be presented as an exact number.
 */
export async function fetchCopilotSpend(
  window: DateWindow,
  apiKey: string,
  org: string,
  seatPriceUsd?: number,
): Promise<AdapterResult> {
  if (seatPriceUsd !== undefined && !(seatPriceUsd >= 0)) {
    throw new InvalidCliArgError(
      `copilot seat price must be a non-negative number, got ${seatPriceUsd}`,
    );
  }

  if (window.start < COPILOT_METRICS_START_DATE) {
    throw new DataUnavailableError(
      TOOL,
      `requested window starts ${window.start}, before Copilot usage metrics reports' ${COPILOT_METRICS_START_DATE} start date`,
    );
  }

  const authHeader = {
    Authorization: `Bearer ${apiKey}`,
    Accept: "application/vnd.github+json",
    "X-GitHub-Api-Version": COPILOT_API_VERSION,
  };

  const userTotals = new Map<string, UserUsage>();

  for (const day of splitIntoDays(window)) {
    const url = `https://api.github.com/orgs/${encodeURIComponent(org)}/copilot/metrics/reports/users-1-day?day=${day}`;
    const report = (await fetchWithRetry({
      tool: TOOL,
      url,
      authHeader,
      emptyOn: [404],
    })) as CopilotReportIndex | null;

    // 404 means "no report for this day" (no Copilot activity, or the org's
    // metrics collection hadn't started yet on that specific day) -- treated
    // as zero users for that day, not a failure, the same way Cursor's
    // adapter treats a 200 { users: [] } response.
    if (report === null) continue;

    const asRecord = report as unknown as Record<string, unknown>;
    const downloadLinks = requireField<string[]>(
      asRecord,
      "download_links",
      TOOL,
    );

    for (const link of downloadLinks) {
      let ndjson: string;
      try {
        // Download links are short-lived, pre-signed GitHub-owned URLs --
        // they carry their own signature/auth in the URL itself and don't
        // need (and can reject) the org's TEAMSPEND_COPILOT_TOKEN.
        ndjson = (await fetchWithRetry({
          tool: TOOL,
          url: link,
          authHeader: {},
          responseType: "text",
        })) as string;
      } catch (error) {
        if (error instanceof AuthenticationError) {
          throw new Error(
            `${TOOL}: report download link for ${day} was rejected (expired signed URL?) -- retry the command (org=${org})`,
          );
        }
        throw error;
      }

      for (const rawUser of parseNdjson(ndjson)) {
        const normalized = normalizeUser(rawUser);
        const existing = userTotals.get(normalized.userId);
        if (existing) {
          existing.requests = (existing.requests ?? 0) + (normalized.requests ?? 0);
          existing.costUsd += normalized.costUsd;
        } else {
          userTotals.set(normalized.userId, normalized);
        }
      }
    }
  }

  if (seatPriceUsd) {
    for (const user of userTotals.values()) {
      user.costUsd += seatPriceUsd;
    }
  }

  const users = [...userTotals.values()];
  return {
    source: TOOL,
    window,
    totalCostUsd: sumCost(users),
    isEstimated: true,
    users,
  };
}

File: python/src/teamspend/adapters/copilot.py
"""
GitHub Copilot usage metrics adapter.

Ported from src/adapters/copilot.ts. See that file's module docstring for
the full research writeup; summarized here:

- The endpoint this adapter calls, GET /orgs/{org}/copilot/metrics/reports/
  users-1-day, is the CURRENT (non-deprecated) Copilot usage metrics API.
  The older /orgs/{org}/copilot/metrics endpoint, which returned inline
  org-aggregate JSON, was sunset by GitHub on 2026-04-02 and no longer
  exists.
- That real API has no arbitrary start/end range parameter -- only
  single-day (`users-1-day?day=YYYY-MM-DD`) or latest-rolling-28-day
  (`users-28-day/latest`) granularity. This adapter requests one report per
  calendar day in the window and sums per-user totals across days, the
  Python mirror of fetch_cursor_spend's 30-day-chunk-and-sum pattern.
- Each report call returns `{download_links, report_day}`, not data inline.
  download_links are short-lived, pre-signed GitHub-owned URLs pointing to
  NDJSON (newline-delimited JSON) files -- this adapter fetches and parses
  each one.
- GitHub's Copilot metrics API has NO cost/spend field at all (unlike
  Cursor/Claude Code, which do report a native cost figure). The only real,
  vendor-reported, per-user number is `ai_credits_used`, which this adapter
  converts to USD at GitHub's own published, fixed $0.01/credit rate --
  never fabricated. An optional `seat_price_usd` adds a flat per-seat
  monthly price once per user (never per day), since GitHub does not expose
  an org's actual negotiated seat price via any API -- the same structural
  gap Cursor's and Claude Code's flat-seat plans have.
- `is_estimated` is unconditionally True for every Copilot result: every
  dollar figure produced here is derived, never vendor-reported.
"""
from __future__ import annotations

import json as json_module
from datetime import date, timedelta
from typing import Any, Dict, List, Optional
from urllib.parse import quote

from ..errors import AuthenticationError, DataUnavailableError, InvalidCliArgError
from ..http_client import fetch_with_retry, require_field
from ..types import AdapterResult, DateWindow, UserUsage, sum_cost

TOOL = "copilot"
COPILOT_API_VERSION = "2026-03-10"

COPILOT_METRICS_START_DATE = "2025-10-10"
"""GitHub's Copilot usage metrics reports have no data before this date."""

COPILOT_CREDIT_USD_RATE = 0.01
"""
GitHub's own published, fixed conversion rate from an AI credit to USD --
not a negotiated or per-org price. Source: GitHub Blog, "GitHub Copilot is
moving to usage-based billing" (2026) -- "1 AI credit = $0.01 USD".
"""


def _split_into_days(window: DateWindow) -> List[str]:
    start = date.fromisoformat(window.start)
    end = date.fromisoformat(window.end)
    days: List[str] = []
    cursor = start
    while cursor <= end:
        days.append(cursor.isoformat())
        cursor += timedelta(days=1)
    return days


def _normalize_user(raw: Dict[str, Any]) -> UserUsage:
    user_login = require_field(raw, "user_login", TOOL)
    ai_credits_used = require_field(raw, "ai_credits_used", TOOL)
    interaction_count = require_field(raw, "user_initiated_interaction_count", TOOL)

    return UserUsage(
        user_id=user_login,
        # GitHub identifies Copilot users by user_id/user_login (a GitHub
        # username), never an email address -- unlike Cursor and Claude Code.
        user_email=None,
        # Copilot's metrics reports carry no token counts at all.
        input_tokens=None,
        output_tokens=None,
        cache_read_tokens=None,
        cache_write_tokens=None,
        requests=interaction_count,
        cost_usd=ai_credits_used * COPILOT_CREDIT_USD_RATE,
        is_estimated=True,
    )


def _parse_ndjson(text: str) -> List[Dict[str, Any]]:
    return [
        json_module.loads(line)
        for line in text.split("\n")
        if line.strip()
    ]


def fetch_copilot_spend(
    window: DateWindow,
    api_key: str,
    org: str,
    seat_price_usd: Optional[float] = None,
    **fetch_kwargs: Any,
) -> AdapterResult:
    """
    Fetches GitHub Copilot org usage for the given window and converts it to
    a dollar figure. See this module's docstring for the full API research
    writeup.

    `fetch_kwargs` forwards to http_client.fetch_with_retry (e.g.
    `transport` and `sleep` for tests) -- never used by real callers.
    """
    if seat_price_usd is not None and not (seat_price_usd >= 0):
        raise InvalidCliArgError(
            f"copilot seat price must be a non-negative number, got {seat_price_usd}"
        )

    if window.start < COPILOT_METRICS_START_DATE:
        raise DataUnavailableError(
            TOOL,
            f"requested window starts {window.start}, before Copilot usage "
            f"metrics reports' {COPILOT_METRICS_START_DATE} start date",
        )

    auth_header = {
        "Authorization": f"Bearer {api_key}",
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": COPILOT_API_VERSION,
    }

    user_totals: Dict[str, UserUsage] = {}

    for day in _split_into_days(window):
        url = (
            f"https://api.github.com/orgs/{quote(org, safe='')}"
            f"/copilot/metrics/reports/users-1-day?day={day}"
        )
        report = fetch_with_retry(
            TOOL, url, auth_header, empty_on=(404,), **fetch_kwargs
        )

        # 404 means "no report for this day" -- treated as zero users for
        # that day, not a failure, the same way Cursor's adapter treats a
        # 200 {"users": []} response.
        if report is None:
            continue

        download_links = require_field(report, "download_links", TOOL)

        for link in download_links:
            try:
                # Download links are short-lived, pre-signed GitHub-owned
                # URLs -- they don't need (and can reject) the org's
                # TEAMSPEND_COPILOT_TOKEN.
                ndjson_text = fetch_with_retry(
                    TOOL,
                    link,
                    {},
                    response_type="text",
                    **fetch_kwargs,
                )
            except AuthenticationError as error:
                raise RuntimeError(
                    f"{TOOL}: report download link for {day} was rejected "
                    f"(expired signed URL?) -- retry the command (org={org})"
                ) from error

            for raw_user in _parse_ndjson(ndjson_text):
                normalized = _normalize_user(raw_user)
                existing = user_totals.get(normalized.user_id)
                if existing:
                    existing.requests = (existing.requests or 0) + (
                        normalized.requests or 0
                    )
                    existing.cost_usd += normalized.cost_usd
                else:
                    user_totals[normalized.user_id] = normalized

    if seat_price_usd:
        for user in user_totals.values():
            user.cost_usd += seat_price_usd

    users_list = list(user_totals.values())
    return AdapterResult(
        source=TOOL,
        window=window,
        total_cost_usd=sum_cost(users_list),
        is_estimated=True,
        users=users_list,
    )
